src/oob/idcert.rs

```diff
@@ -0,0 +1,526 @@
+    pub fn validate_ee(
+        self,
+        issuer: &IdCert,
+    ) -> Result<Self, ValidationError> {
+        self.validate_basics()?;
+        self.validate_issued(issuer)?;
+
+        // Basic Constraints: Must not be present.
+        if self.extensions.basic_ca != None {
+            return Err(ValidationError)
+        }
+
+        // Verify that this is signed by the issuer
+        self.validate_signature(issuer)?;
+        Ok(self)
+    }
+
+
+    //--- Validation Components
+
+    /// Validates basic compliance with RFC8183 and RFC6492
+    ///
+    /// Note the the standards are pretty permissive in this context.
+    fn validate_basics(&self) -> Result<(), ValidationError> {
+        // Validity. Check according to RFC 5280.
+        self.validity.validate()?;
+
+        // Subject Key Identifer. Must be the SHA-1 hash of the octets
+        // of the subjectPublicKey.
+        if self.extensions.subject_key_id.as_slice().unwrap()
+            != self.subject_public_key_info().key_identifier().as_ref()
+        {
+            return Err(ValidationError)
+        }
+
+        Ok(())
+    }
+
+    /// Validates that the certificate is a correctly issued certificate.
+    ///
+    /// Note this check is used to check that an EE certificate in an RFC8183,
+    /// or RFC6492 message is validly signed by the TA certificate that was
+    /// exchanged.
+    ///
+    /// This check assumes for now that we are always dealing with V3
+    /// certificates and AKI and SKI have to match.
+    fn validate_issued(
+        &self,
+        issuer: &IdCert,
+    ) -> Result<(), ValidationError> {
+        // Authority Key Identifier. Must be present and match the
+        // subject key ID of `issuer`.
+        if let Some(ref aki) = self.extensions.authority_key_id {
+            if *aki != issuer.extensions.subject_key_id {
+                return Err(ValidationError)
+            }
+        }
+        else {
+            return Err(ValidationError);
+        }
+
+        Ok(())
+    }
+
+    /// Validates that the certificate is a valid CA certificate.
+    ///
+    /// Checks the parts that are common in normal and trust anchor CA
+    /// certificates.
+    fn validate_ca_basics(&self) -> Result<(), ValidationError> {
+        // 4.8.1. Basic Constraints: For a CA it must be present (RFC6487)
+        // und the “cA” flag must be set (RFC5280).
+        if self.extensions.basic_ca != Some(true) {
+            return Err(ValidationError)
+        }
+
+        Ok(())
+    }
+
+    /// Validates the certificate’s signature.
+    fn validate_signature(
+        &self,
+        issuer: &IdCert
+    ) -> Result<(), ValidationError> {
+        self.signed_data.verify_signature(issuer.public_key())
+    }
+}
+
+
+//--- AsRef
+
+impl AsRef<IdCert> for IdCert {
+    fn as_ref(&self) -> &Self {
+        self
+    }
+}
+
+//------------ Validity ------------------------------------------------------
+
+#[derive(Clone, Debug)]
+pub struct Validity {
```

Comment:
Perhaps move to ::x509 and replace occurrence in ::cert::Cert?

Also for SubjectPublicKeyInfo below …

Reply:
Now importing from ::cert::Cert instead.. not sure yet whether this should be moved to ::x509.